lib/tom_queue/worker_supervisor.rb

```diff
@@ -0,0 +1,84 @@
+module TomQueue
+  class WorkerSupervisor
+
+    attr_accessor :processes, :before_fork, :after_fork
+
+    def initialize
+      @processes = {}
+      @before_fork = -> { }
+      @after_fork = -> { }
+    end
+
+    def supervise(as:, count: 1, &block)
+      count.times do |i|
+        processes["#{as}#{i + 1}"] = block
+      end
+    end
+
+    def run
+      @stop = false
+      process_ids = {}
+      before_fork.call
+      loop do
+        processes_to_start = (processes.keys - process_ids.values)
+
+        processes_to_start.each do |name|
+          $stdout.puts "Starting '#{name}'..."
+          process_id = fork_process(name)
+          setup_signal_handling_for_process(process_id)
+          process_ids[process_id] = name
+        end
+
+        stopped_process_id, status = wait_for_child_process_to_exit
+
+        if stopped_process_name = process_ids.delete(stopped_process_id)
+          $stderr.puts "#{stopped_process_name} terminated (#{stopped_process_id}) with status #{status.exitstatus}"
+        end
+
+        throttle_loop
+        break if @stop
+      end
+    end
+
+    private
+
+    attr_accessor :stop
+
+    def setup_signal_handling_for_process(process_id)
+      Signal.trap("SIGTERM") do
```

Comment:
Unless I've missed something, this will only kill the last process that was forked out - might be worth adding a quick test...

Signal.trap will replace the existing handler, rather than compose - so you'll only ever have one handler running. It might be worth using it to just place a marker to exit the main loop (I believe waitpid, et.al are interrupted if a signal is received anyway).

Then you could just have a common bit of exit code in the main loop method that sends the term process to all processes that it knows about ?